keps/sig-network/3705-cloud-node-ips/README.md

```diff
@@ -0,0 +1,638 @@
+IPv6 node IP but I don't care which one" in cloud-provider clusters
+like they can in non-cloud-provider clusters, but for compatibility
+reasons, we can't change the existing behavior of "`--cloud-provider
+external --node-ip ::`" (which doesn't do what it's "supposed to", but
+does have possibly-useful side effects that have led some users to use
+it anyway; see [kubernetes #111695].)
+
+So instead, we will add new syntax, and allow administrators to say
+"`--node-ip IPv4`" or "`--node-ip IPv6`" if they want to explicitly
+require that the cloud provider pick a node IP of a specific family.
+(This also improves on the behavior of the existing "`0.0.0.0`" and
+"`::`" options, because you almost never actually want the "or fall
+back to the other family if there are no IPs of this family" behavior
+that "`0.0.0.0`" and "`::`" imply.)
+
+```
+<<[UNRESOLVED IPvX capitalization ]>>
+
+Should it be `IPv4`/`IPv6` or `ipv4`/`ipv6`?
+
+<<[/UNRESOLVED]>>
+```
+
+Additionally, we will update the code to allow including "`IPv4`" and
+"`IPv6`" in dual-stack `--node-ip` values as well (in both cloud and
+non-cloud clusters).
+
+[kubernetes #111695]: https://github.com/kubernetes/kubernetes/issues/111695
+
+### Changes to the `provided-node-ip` annotation
+
+Currently, if the user passes an IP address to `--node-ip` which is
+not recognized by the cloud provider as being a valid IP for that
+node, kubelet will set that value in the `provided-node-ip`
+annotation, and the cloud provider will see it, realize that the node
+IP request can't be fulfilled, log an error, and leave the node in the
+tainted state.
+
+It makes sense to have the same behavior if the user passes a "new"
+(eg, dual-stack) `--node-ip` value to kubelet, but the cloud provider
+does not recognize the new syntax and thus can't fulfil the request.
+Conveniently, we can do this just by passing the dual-stack
+`--node-ip` value in the existing annotation; the old cloud provider
+will try to parse it as a single IP address, fail, log an error, and
+leave the node in the tainted state, which is exactly what we wanted
+it to do if it can't interpret the `--node-ip` value correctly.
+
+Therefore, we do not need a new annotation for the new `--node-ip`
+values; we can continue to use the existing annotation, assuming
+existing cloud providers will treat unrecognized values as errors.
+
+```
+<<[UNRESOLVED annotation-name ]>>
+
+The annotation name is `alpha.kubernetes.io/provided-node-ip`
+but it hasn't been "alpha" for a long time. Should we rename it? In
+that case, we probably need to keep supporting both versions for a
+while.
+
+<<[/UNRESOLVED]>>
+```
+
+Kubelet will preserve the existing behavior of _not_ passing
+"`0.0.0.0`" or "`::`" to the cloud provider, for backward
+compatibility, but it _will_ pass "`IPv4`" and "`IPv6`" if they are
+passed as the `--node-ip`.
+
+### Changes to cloud providers
+
+Assuming that all cloud providers use the `"k8s.io/cloud-provider"`
```

Comment:
does this need feature gates there?

Reply:
Hm... I suppose we could make the functions take a boolean saying whether to allow the new syntax or not, or else have a new parsing function for the new syntax... That seems ugly, but I guess if we _don't_ do that then there's no point in even having the cloud provider have a feature gate...

If we rename the annotation to get rid of the "alpha" then we definitely need to be more careful about feature gates, to avoid looking at a stale annotation after a rollback.